Request execution queue

davidwengier opened the pull request:
Part of #46051

Creates a request execution queue that handles the processing of LSP requests by tracking a solution state and ensuring requests that need to mutate that solution state will block future requests until they are finished.

Still to come in later PRs:
* [ ] Handlers to provide a way to indicate which document they need to operate on
* [ ] Remove SolutionProvider from AbstractRequestHandler to force everything to use the RequestContext.Solution
* [ ] Add Document property to RequestContext 
* Actual solution merging will come with #45427

davidfowl commented:
AsyncQueue can be replaced with Channel

davidwengier commented on src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs:
```diff
@@ -0,0 +1,127 @@
+                    // still need to react to it here by throwing away solution updates
+                    if (ranToCompletion)
+                    {
+                        _lastMutatedSolution = mutatedSolution ?? _lastMutatedSolution;
```
There are really 3 options for what to do if a mutating request fails:
1. Throw away the mutations and hope for the best (which is what this code is doing)
2. Apply the mutation and hope for the best
3. Environment.FailFast

I don't have a strong feeling either way really, as I think 1 and 2 both have an equal chance of resulting in a solution state in memory that doesn't match what the user is seeing, but I went with 1 because I am working on the theory that people will be more likely to understand if the server seems to be "lagging" the state they're seeing, and try reconnecting or re-opening the document.

Note that the user experience for errors is unchanged in this PR, so they'll still see the yellow bar etc. and know that something went wrong.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/ExportLspMethodAttribute.cs:
```diff
@@ -15,9 +17,11 @@ internal class ExportLspMethodAttribute : ExportAttribute, IRequestHandlerMetada
+        public bool MutatesSolutionState { get; }
 
-        public ExportLspMethodAttribute(string methodName, string languageName = null) : base(typeof(IRequestHandler))
+        public ExportLspMethodAttribute(string methodName, string? languageName = null, bool mutatesSolutionState = false) : base(typeof(IRequestHandler))
```
i would not make this optional.  all features should have to think about this and be explicit and self documenting.

davidwengier replied:
Agreed, will be in a future PR

CyrusNajmabadi replied:
hrmm... can it be non-optional *in this PR*?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/IRequestHandlerMetadata.cs:
```diff
@@ -14,6 +16,11 @@ internal interface IRequestHandlerMetadata
+        string? LanguageName { get; }
+
+        /// <summary>
+        /// Whether or not handling this method results in changes to the current solution state
```
i would add more information here.  i.e. what the implications of this are.  i.e. that mutation operations will be processed in a serialized fashion.  and non-mutating requests will operate on the last result of the prior mutating requset.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestContext.cs:
```diff
@@ -25,9 +33,25 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
         {
+        }
+
+        internal RequestContext(Solution? solution, Action<Solution>? solutionUpdater, ClientCapabilities clientCapabilities, string? clientName)
```
optional stuff before non-optional seems weird.

davidwengier replied:
Will clean up in a future PR, there is more to add here

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestContext.cs:
```diff
@@ -25,9 +33,25 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+        /// </summary>
+        public void UpdateSolution(Solution solution)
+        {
+            Debug.Assert(_solutionUpdater != null, "This request doesn't allow mutation, so any updates will be ignored by future requests. Do you need to change your ExportLspMethod attribute?");
```
Contract (this should hard throw).

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.QueueItem.cs:
```diff
@@ -0,0 +1,31 @@
+            public Func<RequestContext, Task<bool>> Callback { get; }
+            public bool MutatesSolutionState { get; }
+            public string? ClientName { get; }
+            public ClientCapabilities ClientCapabilities { get; }
```
nit.  up to you.  but i find meaningless properties odd.  FOr basic internal structs, i woudl just make these readonly fields.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs:
```diff
@@ -0,0 +1,127 @@
+    {
+        private readonly AsyncQueue<QueueItem> _queue = new AsyncQueue<QueueItem>();
+        private readonly ILspSolutionProvider _solutionProvider;
+        private Solution? _lastMutatedSolution;
```
doc threading invariants.

davidwengier replied:
I'm not sure what you mean by this. There is no concurrency here because ProcessRequestQueue is the only thing that accesses this. Or do you just mean doc that?

davidwengier replied:
Ended up moving this into the ProcessQueue method

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs:
```diff
@@ -0,0 +1,127 @@
+        /// <param name="cancellationToken">A cancellation token that will cancel the handing of this request.</param>
+        /// <returns>A task that can be awaited to observe the results of the handing of this request.</returns>
+        public Task<TResponseType> ExecuteAsync<TRequestType, TResponseType>(bool mutatesSolutionState, IRequestHandler<TRequestType, TResponseType> handler, TRequestType request,
+            ClientCapabilities clientCapabilities, string? clientName, CancellationToken cancellationToken) where TRequestType : class
```
wrap all params.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs:
```diff
@@ -0,0 +1,127 @@
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        completion.SetCanceled();
+                        return false;
```
doc what the true/false value is for.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs:
```diff
@@ -0,0 +1,127 @@
+                    catch (Exception exception)
+                    {
+                        // Pass the exception to the task completion source, so the caller of the ExecuteAsync method can observe but
+                        // don't let it escape from this callback, so it doens't affect the queue processing.
```
```suggestion
                        // don't let it escape from this callback, so it doesn't affect the queue processing.
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs:
```diff
@@ -0,0 +1,127 @@
+            _solutionProvider = solutionProvider;
+
+            // Start the queue processing
+            _ = ProcessQueueAsync();
```
this should be tied to somethign that shutsdown the queue processing.  Sam has an interface fo rthis i believe.

dibarbet replied:
Perhaps this could be tied to the LSP shutdown messages.  Just need to make sure that it can be started again if the LSP client decides to restart.

davidwengier replied:
I've added a cancellation token for this for now, will add the rest to the TODO list as the lifetime of this class is in question anyway.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs:
```diff
@@ -0,0 +1,127 @@
+                var work = await _queue.DequeueAsync().ConfigureAwait(false);
+
+                var solution = GetCurrentSolution();
+                solution = MergeChanges(solution, _lastMutatedSolution);
```
doc what's going on here.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs:
```diff
@@ -0,0 +1,127 @@
+                    var ranToCompletion = await work.Callback(context).ConfigureAwait(false);
+
+                    // If the handling of the request failed, the exception will bubble back up to the caller, but we
+                    // still need to react to it here by throwing away solution updates
```
need to think about this.  i thinki we should chat on what semantics we want ehre.  

1. if a mutation operation fails, i honestly don't know *waht* we can do.  because our state is literally out sync with whatever host we have.  contiuing will almost certainly lead to crashes.  i think we should just yellow-bar and shutdown there.

davidwengier replied:
The user will still see the yellow bar. I talked about this in a comment here: https://github.com/dotnet/roslyn/pull/46660#discussion_r467659997

What exactly to you mean by "shutdown"? I would want the queue to keep processing because closing and reopening the document should be all that is required to get back to a good state.

CyrusNajmabadi replied:
> I would want the queue to keep processing because closing and reopening the document should be all that is required to get back to a good state.

1. do users know that?
2. in the meantime, we're out of sync, and i wouldn't expect to be able to handle *any* messages for that doc.  i.e. even a simple "what are the colors" request will likely lead to gibberish and/or crashes due to different bounds on the client/server.

davidwengier replied:
I don't disagree, but not not sure what we can do about it. The client could decide to respond to the error by closing/reopening the document (or re-initializing it) and then there wouldn't be any future problem. Maybe we could keep a "bad document list" and refuse to respond to requests for that document until we get a `didOpen`?

We should discuss more formally. This PR is to a feature branch so easy to follow up with changes later.

dibarbet replied:
I can't link the exact location in the spec, but if you search of inconsistent you should find it - https://microsoft.github.io/language-server-protocol/specification

> if servers end up in an inconsistent state they should log this to the client using the window/logMessage request. If they can’t recover from this the best they can do right now is to exit themselves. We are considering an extension to the protocol that allows servers to request a restart on the client side.

It sounds like we could exit, but if the LSP client doesn't attempt to restart us I'm not sure that's the best approach either.

CyrusNajmabadi replied:
> We should discuss more formally.

Yup!

> Maybe we could keep a "bad document list" and refuse to respond to requests for that document until we get a didOpen?

I like this idea.  Including a way to inform users of this.  i.e. the yellow-bar, or console message can state explicitly that we've experienced a syncing problem with a doc and users can try to fix it by closing/opening.

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs:
```diff
@@ -0,0 +1,127 @@
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    [Export(typeof(RequestExecutionQueue)), Shared]
```
Ah, we'll also have multiple LSP servers long term anyway - the XAML LSP server is based off of this infrastructure, so we definitely need to make this per LSP server

davidwengier replied:
Good timing for this comment, because I was actually thinking this morning of moving the queue to the `AbstractRequestHandler` because with the work I did yesterday to pass the Document into the handler, and remove the SolutionProvider, it feels like a bit of a trap being on the server.

I'll have to think about it some more.

davidwengier replied:
> we want different execution queues per language server

This is currently imported by the language server, so we can just remove the `[Shared]` attribute right?

dibarbet commented on src/Features/LanguageServer/ProtocolUnitTests/Ordering/FailingMutatingRequestHandler.cs:
```diff
@@ -0,0 +1,41 @@
+
+namespace Microsoft.CodeAnalysis.LanguageServer.UnitTests.RequestOrdering
+{
+    [Shared, ExportLspMethod(MethodName, mutatesSolutionState: true)]
```
We should export all of these test handlers with PartNotDiscoverable to ensure it won't automatically be included in places we don't want it.

dibarbet commented on src/Features/LanguageServer/ProtocolUnitTests/Ordering/RequestOrderingTests.cs:
```diff
@@ -0,0 +1,210 @@
+        }
+
+        [Fact]
+        public async Task NonMutatingWaitsForMutating()
```
can we add one that tests the behavior when a non-mutating request is in progress and a new mutating request comes in?

davidwengier commented:
Have updated addressing all feedback, but for the items we discussed. I don't think there is anything blocking this PR being merged into the feature branch, but let me know if you disagree.

Have also updated the TODO list on https://github.com/dotnet/roslyn/issues/46051#issue-658515924

davidwengier commented:
Ping @CyrusNajmabadi @dibarbet, would be good to merge this so the next PR is simpler

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/AbstractRequestHandlerProvider.cs:
```diff
@@ -19,9 +19,13 @@ namespace Microsoft.CodeAnalysis.LanguageServer
 
-        public AbstractRequestHandlerProvider(IEnumerable<Lazy<IRequestHandler, IRequestHandlerMetadata>> requestHandlers, string? languageName = null)
-            => _requestHandlers = CreateMethodToHandlerMap(requestHandlers.Where(rh => rh.Metadata.LanguageName == languageName));
+        public AbstractRequestHandlerProvider(IEnumerable<Lazy<IRequestHandler, IRequestHandlerMetadata>> requestHandlers, RequestExecutionQueue queue, string? languageName = null)
```
what does it mean for langaugename to be null?  we would then filter down to request handlers that don't answer requests for a particular langauge?  this seems.... odd.

i'm also not a fan of internal optiona parameters.  it just means it's hard to tell at the callsite what's going on.

dibarbet replied:
This was a new option added for XAML support.  The old handlers weren't updated when this option was added for XAML
Filed https://github.com/dotnet/roslyn/issues/46880 for updating this

davidwengier replied:
This is also covered by https://github.com/dotnet/roslyn/issues/46195 right? We need to split the handers into C# and VB forms for this too.

dibarbet replied:
right, I forgot about that one 😆 
Should be fixed once we switch to content type

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/ExportLspMethodAttribute.cs:
```diff
@@ -15,9 +17,11 @@ internal class ExportLspMethodAttribute : ExportAttribute, IRequestHandlerMetada
-        public string LanguageName { get; }
+        public string? LanguageName { get; }
+
+        public bool MutatesSolutionState { get; }
```
doc

CyrusNajmabadi replied:
even  <see cref=IRequestHandlerMetadata.MutatesSolutionState>

davidwengier replied:
This is in a follow up PR

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/IRequestHandlerMetadata.cs:
```diff
@@ -14,6 +16,13 @@ internal interface IRequestHandlerMetadata
+
+        /// <summary>
+        /// Whether or not handling this method results in changes to the current solution state.
+        /// Mutating requests will block subsequent requests from starting until after they have
```
can you clkarify if it blocks all subsequent requests, or just other mutating requests.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestContext.cs:
```diff
@@ -25,9 +34,25 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
         public string? ClientName { get; }
 
         public RequestContext(ClientCapabilities clientCapabilities, string? clientName)
+            : this(null, null, clientCapabilities, clientName)
```
1. named params for null.
2. my preference is that the common params be at teh same location.  so it would be `ClientCap, clientName, solution, solutionUpdated)

davidwengier replied:
This is cleaned up in a follow up PR

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestContext.cs:
```diff
@@ -25,9 +34,25 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+        public void UpdateSolution(Solution solution)
+        {
+            Contract.ThrowIfNull(_solutionUpdater, "Mutating solution not allowed in a non-mutating request handler");
+            _solutionUpdater?.Invoke(solution);
```
```suggestion
            _solutionUpdater.Invoke(solution);
```

CyrusNajmabadi replied:
also, this is presumably a one-and-done sort of thing.  it might be good to store if we've applied the func, and throw if called twice.

davidwengier replied:
I thought about this, and I'm not sure I see the benefit of throwing over just taking the last updated solution that we're given. I'm going to leave this decision for later, after we actually have a handler that mutates, and see if which form fits better.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs:
```diff
@@ -0,0 +1,155 @@
+                    // Tell the queue to ignore any mutations from this request
+                    return false;
+                }, requestCancellationToken);
+            _queue.Enqueue(item);
```
is asyncqueue threadsafe?

dibarbet replied:
per the docs it appears to be 
https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.threading.asyncqueue-1?view=visualstudiosdk-2019

davidwengier replied:
Says so on the tin :)
https://github.com/microsoft/vs-threading/blob/master/src/Microsoft.VisualStudio.Threading/AsyncQueue.cs#L19

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.QueueItem.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        private readonly struct QueueItem
+        {
+            public readonly Func<RequestContext, CancellationToken, Task<bool>> Callback;
```
doc what the 'bool' return type of this function is.

CyrusNajmabadi replied:
nit:

```suggestion
            public readonly Func<RequestContext, CancellationToken, Task<bool>> CallbackAsync;
```

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs:
```diff
@@ -0,0 +1,155 @@
+                // The "current" solution can be updated by non-LSP actions, so we need it, but we also need
+                // to merge in the changes from any mutations that have been applied to open documents
+                var solution = GetCurrentSolution();
+                solution = MergeChanges(solution, lastMutatedSolution);
```
so we have like two mutations going on simultaneously that we're trying to reconcile?  i think i hate that.  why is it not hte case that any mutations just affect the single canonical workspace.CurrentSOlution?

dibarbet replied:
I believe this is just referencing the issue where we keep track of LSP open documents separately from the workspace's current solution - https://github.com/dotnet/roslyn/issues/45427

We should only ever have 1 LSP mutation at a time, but the workspace's current solution can change (e.g. adding project references)

davidwengier replied:
@dibarbet is correct. Solution changes can still come from the workspace, and then we'll merge in the changes for the documents that are open via LSP.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs:
```diff
@@ -0,0 +1,155 @@
+                else
+                {
+                    // Non mutating request get given the current solution state, but are otherwise fire-and-forget
+                    _ = work.Callback(context, cancellationToken);
```
fire-and-forget means we lose exceptions.  this seems non-ideal.

dibarbet replied:
agree - should definitely track these in some manner, maybe NFW?

davidwengier replied:
Will do this with the failed mutation handling. Since only an exception inside the callback function (ie, not from the actual request handler) would possibly bubble out of here, shutting down the server is probably the only course of action really.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs:
```diff
@@ -0,0 +1,155 @@
+        {
+            // TODO: Merge in changes to the solution that have been received from didChange LSP methods
+            // https://github.com/dotnet/roslyn/issues/45427
+            return mutatedSolution ?? solution;
```
errr.

davidwengier commented:
Updated the docs per the comments. Going to merge this when CI finishes to push up the next piece.
